From 2e4ad49d23862e6bbb50a1b2cc4a98df45eb94b1 Mon Sep 17 00:00:00 2001
From: Niels de Vos <ndevos@redhat.com>
Date: Wed, 29 Jul 2015 11:20:20 +0200
Subject: [PATCH] build: make it possible to disable FSAL_NULL

When FSAL_NULL gets build, there will be an additional
nfs-ganesha-nullfs RPM generated. To prevent any confusion if this
package is needed or not, add an option to disable the building of
FSAL_NULL and don't build the RPM package in that case.

The usage is similar to other FSALs, FSAL_NULL is enabled by default,
disabling cam be done like this:

  $ cmake -DUSE_FSAL_NULL=NO ... /path/to/nfs-ganesha/src/

Change-Id: Ifc932fd35b195bc1ea33dd2b899b4d1bab0dbab3
Signed-off-by: Niels de Vos <ndevos@redhat.com>
---
 src/CMakeLists.txt                      |  8 ++++++++
 src/FSAL/Stackable_FSALs/CMakeLists.txt |  2 ++
 src/nfs-ganesha.spec-in.cmake           | 36 ++++++++++++++++++++-------------
 3 files changed, 32 insertions(+), 14 deletions(-)

diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 8e9ae94..af7215e 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -177,6 +177,7 @@ option(USE_FSAL_LUSTRE "build LUSTRE FSAL" ON)
 option(USE_FSAL_XFS "build XFS support in VFS FSAL" ON)
 option(USE_FSAL_PANFS "build PanFS support in VFS FSAL" ON)
 option(USE_FSAL_GLUSTER "build GLUSTER FSAL shared library" ON)
+option(USE_FSAL_NULL "build NULL FSAL shared library" ON)
 
 # FSALs which are disabled by default
 option(USE_FSAL_PT "build PT FSAL" OFF)
@@ -953,6 +954,7 @@ message(STATUS "USE_FSAL_LUSTRE = ${USE_FSAL_LUSTRE}")
 message(STATUS "USE_FSAL_SHOOK = ${USE_FSAL_SHOOK}")
 message(STATUS "USE_FSAL_LUSTRE_UP = ${USE_FSAL_LUSTRE_UP}")
 message(STATUS "USE_FSAL_GLUSTER = ${USE_FSAL_GLUSTER}")
+message(STATUS "USE_FSAL_NULL = ${USE_FSAL_NULL}")
 message(STATUS "USE_DBUS = ${USE_DBUS}")
 message(STATUS "USE_CB_SIMULATOR = ${USE_CB_SIMULATOR}")
 message(STATUS "USE_NFSIDMAP = ${USE_NFSIDMAP}")
@@ -1254,6 +1256,12 @@ else(USE_FSAL_PT)
     set(BCOND_PT "%bcond_with")
 endif(USE_FSAL_PT)
 
+if(USE_FSAL_NULL)
+    set(BCOND_NULLFS "%bcond_without")
+else(USE_FSAL_NULL)
+    set(BCOND_NULLFS "%bcond_with")
+endif(USE_FSAL_NULL)
+
 if(USE_9P_RDMA)
     set(BCOND_RDMA "%bcond_without")
 else(USE_9P_RDMA)
diff --git a/src/FSAL/Stackable_FSALs/CMakeLists.txt b/src/FSAL/Stackable_FSALs/CMakeLists.txt
index 470ba7b..1408423 100644
--- a/src/FSAL/Stackable_FSALs/CMakeLists.txt
+++ b/src/FSAL/Stackable_FSALs/CMakeLists.txt
@@ -1 +1,3 @@
+if(USE_FSAL_NULL)
 add_subdirectory(FSAL_NULL)
+endif(USE_FSAL_NULL)
diff --git a/src/nfs-ganesha.spec-in.cmake b/src/nfs-ganesha.spec-in.cmake
index 4943f37..28ae150 100644
--- a/src/nfs-ganesha.spec-in.cmake
+++ b/src/nfs-ganesha.spec-in.cmake
@@ -25,6 +25,9 @@
 # %bcond_with means you add a "--with" option, default = without this feature
 # %bcond_without adds a"--without" so the feature is enabled by default
 
+@BCOND_NULLFS@ nullfs
+%global use_fsal_null %{on_off_switch nullfs}
+
 @BCOND_GPFS@ gpfs
 %global use_fsal_gpfs %{on_off_switch gpfs}
 
@@ -145,15 +148,6 @@ Requires: nfs-ganesha = %{version}-%{release}
 This package contains a FSAL shared object to
 be used with NFS-Ganesha to support VFS based filesystems
 
-%package nullfs
-Summary: The NFS-GANESHA's NULLFS Stackable FSAL
-Group: Applications/System
-Requires: nfs-ganesha = %{version}-%{release}
-
-%description nullfs
-This package contains a Stackable FSAL shared object to
-be used with NFS-Ganesha. This is mostly a template for future (more sophisticated) stackable FSALs
-
 %package proxy
 Summary: The NFS-GANESHA's PROXY FSAL
 Group: Applications/System
@@ -194,6 +188,18 @@ to the ganesha.nfsd server, it makes it possible to trace using LTTng.
 # Option packages start here. use "rpmbuild --with lustre" (or equivalent)
 # for activating this part of the spec file
 
+# NULL
+%if %{with nullfs}
+%package nullfs
+Summary: The NFS-GANESHA's NULLFS Stackable FSAL
+Group: Applications/System
+Requires: nfs-ganesha = %{version}-%{release}
+
+%description nullfs
+This package contains a Stackable FSAL shared object to
+be used with NFS-Ganesha. This is mostly a template for future (more sophisticated) stackable FSALs
+%endif
+
 # GPFS
 %if %{with gpfs}
 %package gpfs
@@ -332,6 +338,7 @@ be used with NFS-Ganesha to support Gluster
 %build
 cmake .	-DCMAKE_BUILD_TYPE=Debug			\
 	-DBUILD_CONFIG=rpmbuild				\
+	-DUSE_FSAL_NULL=%{use_fsal_null}		\
 	-DUSE_FSAL_ZFS=%{use_fsal_zfs}			\
 	-DUSE_FSAL_XFS=%{use_fsal_xfs}			\
 	-DUSE_FSAL_CEPH=%{use_fsal_ceph}		\
@@ -480,16 +487,17 @@ make DESTDIR=%{buildroot} install
 %config(noreplace) %{_sysconfdir}/ganesha/vfs.conf
 
 
-%files nullfs
-%defattr(-,root,root,-)
-%{_libdir}/ganesha/libfsalnull*
-
-
 %files proxy
 %defattr(-,root,root,-)
 %{_libdir}/ganesha/libfsalproxy*
 
 # Optional packages
+%if %{with nullfs}
+%files nullfs
+%defattr(-,root,root,-)
+%{_libdir}/ganesha/libfsalnull*
+%endif
+
 %if %{with gpfs}
 %files gpfs
 %defattr(-,root,root,-)
-- 
2.1.0

